openfermioncirq/gates/fermionic_simulation.py

```diff
@@ -0,0 +1,453 @@
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+from typing import Optional, Tuple, Union
+
+import cirq
+import numpy as np
+import scipy.linalg as la
+import sympy
+
+from openfermioncirq.gates.common_gates import XXYYPowGate
+
+
+def _arg(x):
+    if x == 0:
+        return 0
+    if cirq.is_parameterized(x):
+        return sympy.arg(x)
+    return np.angle(x)
+
+
+def _canonicalize_weight(w):
+    if w == 0:
+        return (0, 0)
+    if cirq.is_parameterized(w):
+        return (cirq.PeriodicValue(abs(w), 4), sympy.arg(w))
+    return (np.round((w % 4) if (w == np.real(w)) else
+        (abs(w) % 4) * w / abs(w), 8), 0)
+
+
+def state_swap_eigen_component(x: str, y: str, sign: int = 1, angle: float=0):
+    """The +/- eigen-component of the operation that swaps states x and y.
+
+    For example, state_swap_eigen_component('01', '10', ±1) with angle θ returns
+        ┌                               ┐
+        │0, 0,           0,            0│
+        │0, 0.5,         ±0.5 e^{-iθ}, 0│
+        │0, ±0.5 e^{iθ}, 0.5,          0│
+        │0, 0,           0,            0│
+        └                               ┘
+
+    Args:
+        x: The first state to swap, as a bitstring.
+        y: The second state to swap, as a bitstring. Must have high index than
+            x.
+        sign: The sign of the off-diagonal elements (indicated by +/-1).
+        angle: The phase of the complex off-diagonal elements. Defaults to 0.
+
+    Returns: The eigen-component.
+
+    Raises:
+        ValueError:
+            * x and y have different lengths
+            * x or y contains a character other than '0' and '1'
+            * x and y are the same
+            * sign is not -1 or 1
+        TypeError: x or y is not a string
+    """
+    if not (isinstance(x, str) and isinstance(y, str)):
+        raise TypeError('not (isinstance(x, str) and isinstance(y, str))')
+    if len(x) != len(y):
+        raise ValueError('len(x) != len(y)')
+    if set(x).union(y).difference('01'):
+        raise ValueError('Arguments must be 0-1 strings.')
+    if x == y:
+        raise ValueError('x == y')
+    if sign not in (-1, 1):
+        raise ValueError('sign not in (-1, 1)')
+
+    dim = 2 ** len(x)
+    i, j = int(x, 2), int(y, 2)
+
+    component = np.zeros((dim, dim), dtype=np.complex128)
+    component[i, i] = component[j, j] = 0.5
+    component[j, i]= sign * 0.5 * 1j**(angle * 2 / np.pi)
+    component[i, j]= sign * 0.5 * 1j**(-angle * 2 / np.pi)
+    return component
+
+
+class QuadraticFermionicSimulationGate(
+        cirq.EigenGate,
+        cirq.InterchangeableQubitsGate,
+        cirq.TwoQubitGate):
+    """(w0 |10><01| + h.c.) + w1 * |11><11| interaction.
+
+    With weights (w0, w1) and exponent t, this gate's matrix is defined as
+    follows:
+        exp(-i t ((w0 |10><01| + h.c.) + w1 |11><11|))
```

Comment:
Please write in the documentation a bit about how these gates arise in fermionic simulations, referencing the ladder operators.

Reply:
Done